### PR TITLE
Support skip_values in DictionaryDecoder

### LINKED ISSUE #2709: Add divide_opt kernel which produce null values on division by zero error
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

To apply arithmetic kernels on large data, we need a kind of divide kernel that can tolerate fatal error like division by zero and produce null instead of failing operation.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Ted-Jiang**
# Which issue does this PR close?

Closes #2709.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/src/arrow/array_reader/byte_array_dictionary.rs [hunk 1]
```diff
@@ -459,7 +481,75 @@ mod tests {
...
+        assert_eq!(decoder.read(&mut output, 0..2).unwrap(), 2);
+
+        let mut valid = vec![false, false, false, true, false, true];
+        let valid_buffer = Buffer::from_iter(valid.iter().cloned());
+        output.pad_nulls(0, 2, valid.len(), valid_buffer.as_slice());
+
+        assert!(matches!(output, DictionaryBuffer::Dict { .. }));
+
+        assert_eq!(decoder.skip_values(1).unwrap(), 1);
+        assert_eq!(decoder.read(&mut output, 0..3).unwrap(), 3);
```
Comment (Ted-Jiang) [inline parquet/src/arrow/array_reader/byte_array_dictionary.rs:521]: @tustvold  Like you metion in https://github.com/apache/arrow-rs/pull/2076#discussion_r922306318
If i try to use ` assert_eq!(decoder.read(&mut output, 3..6).unwrap(), 3);`

The result is not right, also appear in `test_dictionary_preservation`
i think we should align it.

File: parquet/src/arrow/array_reader/byte_array_dictionary.rs [hunk 2]
```diff
@@ -368,16 +368,38 @@ where
                             dict_offsets,
                             dict_values,
                         )?;
-
+                        *max_remaining_values -= len;
```
Comment (Ted-Jiang) [inline parquet/src/arrow/array_reader/byte_array_dictionary.rs:371]: we should update `max_remaining_values` here.

File: parquet/src/arrow/array_reader/byte_array.rs [hunk 3]
```diff
@@ -406,6 +420,7 @@ impl ByteArrayDecoderPlain {
             skip += 1;
             self.offset = self.offset + 4 + len;
         }
+        self.max_remaining_values -= skip;
```
Comment (Ted-Jiang) [inline parquet/src/arrow/array_reader/byte_array.rs:423]: forgot update max_remaining_values 🤣

File: parquet/src/arrow/array_reader/byte_array_dictionary.rs [hunk 4]
```diff
@@ -368,16 +368,38 @@ where
...
+                decoder,
+                max_remaining_values, } => {
+                let to_skip = num_values.min(*max_remaining_values);
+
+                let dict = self
+                    .dict
+                    .as_ref()
+                    .ok_or_else(|| general_err!("missing dictionary page for column"))?;
+
+                if dict.is_empty() {
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/byte_array_dictionary.rs:394]: I'm not sure why this check is necessary, removing the corresponding check from read, doesn't seem to fail any tests

**Review by tustvold — COMMENTED** [review]
I started reviewing this before I realised I'd just merged #2097... How would you like to proceed, I think there might perhaps be value in some of the tests??

File: parquet/src/arrow/array_reader/byte_array_dictionary.rs (hunk 1 above)
Comment (Ted-Jiang) [reply parquet/src/arrow/array_reader/byte_array_dictionary.rs:521]: i don't think so ,  like previous  test read two range  '0..3' and '0..4' 
https://github.com/apache/arrow-rs/blob/a7f6a6e9261f2320ed34500d6d649efc97617161/parquet/src/arrow/array_reader/byte_array_dictionary.rs#L444-L453

 it seems it total read 7 values 
https://github.com/apache/arrow-rs/blob/a7f6a6e9261f2320ed34500d6d649efc97617161/parquet/src/arrow/array_reader/byte_array_dictionary.rs#L468-L486

**Comment by Ted-Jiang** [comment]
duplicate with https://github.com/apache/arrow-rs/pull/2100

**Pull request closed without merging**